### ISSUE TITLE
Scheduled Jobs No Longer Running - Status "Stopped"

### ISSUE THREAD

**Issue by TallGibbs** [jupyter-server/jupyter-scheduler#529]
## Description
I made a silly mistake on June 12th, and updated multiple python packages all at the same time (from my Anaconda Navigator).  Ever since then, all of my scheduled jobs in Jupyter Lab have Status of "Stopped" and I cannot get any scheduled job to run on a schedule.  However, I can open the notebook directly and run all cells perfectly.  So, the notebook code still works fine for all of the jobs; it is only an issue with the schedule jobs.

## Reproduce

1. Go to a notebook file you want to run
2. Right Click on the file and hit "Create Notebook Job"
3. On the "Create Job" tab that opens, just do "Run now" (or you can schedule a job for a later time to test that as well).  
4. See error "Unexpected error occurred during creation of job" occurs immediately and the job cannot run.

## Expected behavior

The jobs previously were scheduled and ran fine on their own.  Also the "Run now" button worked just fine to start the job.

## Context

- Operating System and version: Windows 10
- Browser and version: Chrome
- Jupyter Server version: will include in the "jupyter troubleshoot" result in attachment

<details><summary>Troubleshoot Output</summary>
<pre>
C:\>jupyter troubleshoot
$PATH:
        C:\Users\XXXXX\AppData\Local\anaconda3\envs\Projects
        C:\Users\XXXXX\AppData\Local\anaconda3\envs\Projects\Library\mingw-w64\bin
        C:\Users\XXXXX\AppData\Local\anaconda3\envs\Projects\Library\usr\bin
        C:\Users\XXXXX\AppData\Local\anaconda3\envs\Projects\Library\bin
        C:\Users\XXXXX\AppData\Local\anaconda3\envs\Projects\Scripts
        C:\Users\XXXXX\AppData\Local\anaconda3\envs\Projects\bin
        C:\Users\XXXXX\AppData\Local\anaconda3\condabin
        C:\WINDOWS\system32
        C:\WINDOWS
        C:\WINDOWS\System32\Wbem
        C:\WINDOWS\System32\WindowsPowerShell\v1.0
        C:\WINDOWS\System32\OpenSSH
        C:\Program Files\Citrix\System32
        C:\Program Files\Citrix\ICAService
        C:\app\client\32\product\19.0.0\client_1
        C:\app\client\64\product\19.0.0\client_1\bin
        C:\Program Files\PuTTY
        C:\Program Files (x86)\Webex\Plugins
        C:\ProgramData\Oracle\Java\javapath
        C:\Program Files (x86)\Microsoft SQL Server\150\DTS\Binn
        C:\Program Files\Azure Data Studio\bin
        C:\Program Files\Python310\Scripts
        C:\Program Files\Python310
        C:\Program Files (x86)\Aperio\Common
        C:\Program Files\dotnet
        C:\Program Files (x86)\Citrix\HDX\bin
        C:\Program Files\Citrix\HDX\bin
        C:\Users\XXXXX\AppData\Local\Microsoft\WindowsApps
        .
        C:\Program Files\Azure Data Studio\bin
        C:\Users\XXXXX\.dotnet\tools
        C:\Users\XXXXX\AppData\Local\Programs\Microsoft VS Code\bin

sys.path:
        C:\Users\XXXXX\AppData\Local\anaconda3\envs\Projects\Scripts
        C:\Users\XXXXX\AppData\Local\anaconda3\envs\Projects\python311.zip
        C:\Users\XXXXX\AppData\Local\anaconda3\envs\Projects\DLLs
        C:\Users\XXXXX\AppData\Local\anaconda3\envs\Projects\Lib
        C:\Users\XXXXX\AppData\Local\anaconda3\envs\Projects
        C:\Users\XXXXX\AppData\Roaming\Python\Python311\site-packages
        C:\Users\XXXXX\AppData\Roaming\Python\Python311\site-packages\win32
        C:\Users\XXXXX\AppData\Roaming\Python\Python311\site-packages\win32\lib
        C:\Users\XXXXX\AppData\Roaming\Python\Python311\site-packages\Pythonwin
        C:\Users\XXXXX\AppData\Local\anaconda3\envs\Projects\Lib\site-packages
        C:\Users\XXXXX\AppData\Local\anaconda3\envs\Projects\Lib\site-packages\win32
        C:\Users\XXXXX\AppData\Local\anaconda3\envs\Projects\Lib\site-packages\win32\lib
        C:\Users\XXXXX\AppData\Local\anaconda3\envs\Projects\Lib\site-packages\Pythonwin

sys.executable:
        C:\Users\XXXXX\AppData\Local\anaconda3\envs\ Projects\python.exe

sys.version:
        3.11.9 | packaged by Anaconda, Inc. | (main, Apr 19 2024, 16:40:41) [MSC v.1916 64 bit (AMD64)]

platform.platform():
        Windows-10-10.0.19045-SP0

where jupyter:
        C:\Users\XXXXX\AppData\Local\anaconda3\envs\Projects\Scripts\jupyter.exe

pip list:
        Package                   Version
        ------------------------- ---------------
        aiohttp                   3.9.3
        aiosignal                 1.3.1
        annotated-types           0.6.0
        anyio                     4.2.0
        argon2-cffi               21.3.0
        argon2-cffi-bindings      21.2.0
        asttokens                 2.0.5
        async-lru                 2.0.4
        attrs                     23.1.0
        Babel                     2.11.0
        bcrypt                    4.1.2
        beautifulsoup4            4.12.3
        bleach                    4.1.0
        Brotli                    1.0.9
        cattrs                    23.2.3
        certifi                   2024.6.2
        cffi                      1.16.0
        charset-normalizer        2.0.4
        colorama                  0.4.6
        comm                      0.2.1
        contourpy                 1.2.0
        croniter                  1.4.1
        cryptography              42.0.4
        cycler                    0.11.0
        daal4py                   2023.1.1
        dataclasses-json          0.6.4
        debugpy                   1.6.7
        decorator                 5.1.1
        defusedxml                0.7.1
        distlib                   0.3.8
        distro                    1.9.0
        et-xmlfile                1.1.0
        executing                 0.8.3
        fastapi                   0.103.0
        fastf1                    3.2.1
        fastjsonschema            2.16.2
        filelock                  3.13.1
        fonttools                 4.51.0
        fqdn                      1.5.1
        frozenlist                1.4.1
        fsspec                    2023.6.0
        greenlet                  3.0.3
        h11                       0.14.0
        httpcore                  1.0.2
        httpx                     0.26.0
        idna                      3.7
        imagecodecs               2023.1.23
        imageio                   2.33.1
        ipykernel                 6.28.0
        ipython                   8.25.0
        isoduration               20.11.0
        jedi                      0.18.1
        Jinja2                    3.1.4
        joblib                    1.4.2
        json5                     0.9.6
        jsonpatch                 1.33
        jsonpointer               2.4
        jsonschema                4.19.2
        jsonschema-specifications 2023.7.1
        jupyter_client            8.6.0
        jupyter_core              5.7.2
        jupyter-events            0.10.0
        jupyter-lsp               2.2.0
        jupyter_scheduler         2.4.0
        jupyter_server            2.14.1
        jupyter_server_terminals  0.4.4
        jupyterlab                4.0.11
        jupyterlab-pygments       0.1.2
        jupyterlab_server         2.25.1
        kiwisolver                1.4.4
        langchain                 0.1.13
        langchain-community       0.0.29
        langchain-core            0.1.33
        langchain-experimental    0.0.55
        langchain-openai          0.1.0
        langchain-text-splitters  0.0.1
        langsmith                 0.1.31
        lazy_loader               0.4
        MarkupSafe                2.1.3
        marshmallow               3.20.2
        matplotlib                3.8.4
        matplotlib-inline         0.1.6
        mistune                   2.0.4
        mkl-fft                   1.3.8
        mkl-random                1.2.4
        mkl-service               2.4.0
        multidict                 6.0.5
        mypy-extensions           1.0.0
        nbclient                  0.8.0
        nbconvert                 7.10.0
        nbformat                  5.9.2
        nest-asyncio              1.6.0
        networkx                  3.2.1
        notebook                  7.0.8
        notebook_shim             0.2.3
        numpy                     1.26.3
        openai                    1.12.0
        openpyxl                  3.1.2
        orjson                    3.9.15
        overrides                 7.4.0
        packaging                 23.2
        pandas                    2.1.4
        pandocfilters             1.5.0
        paramiko                  3.4.0
        parso                     0.8.3
        pillow                    10.3.0
        pip                       24.0
        platformdirs              3.10.0
        prometheus-client         0.14.1
        prompt-toolkit            3.0.43
        psutil                    5.9.0
        pure-eval                 0.2.2
        pycparser                 2.21
        pydantic                  2.5.3
        pydantic_core             2.14.6
        Pygments                  2.15.1
        PyNaCl                    1.5.0
        pyodbc                    5.0.1
        pyparsing                 3.0.9
        PySocks                   1.7.1
        python-dateutil           2.9.0.post0
        python-dotenv             1.0.1
        python-json-logger        2.0.7
        pytz                      2023.3
        pywin32                   306
        pywinpty                  2.0.10
        PyYAML                    6.0.1
        pyzmq                     25.1.2
        rapidfuzz                 3.6.1
        referencing               0.30.2
        regex                     2023.12.25
        requests                  2.32.2
        requests-cache            1.1.1
        rfc3339-validator         0.1.4
        rfc3986-validator         0.1.1
        rpds-py                   0.10.6
        scikit-image              0.23.2
        scikit-learn              1.4.2
        scikit-learn-intelex      20230426.121932
        scikit-plot               0.3.7
        scipy                     1.13.1
        Send2Trash                1.8.2
        setuptools                69.5.1
        six                       1.16.0
        sniffio                   1.3.0
        soupsieve                 2.5
        SQLAlchemy                1.4.51
        stack-data                0.2.0
        starlette                 0.27.0
        tenacity                  8.2.3
        terminado                 0.17.1
        textdistance              4.6.1
        thefuzz                   0.22.1
        threadpoolctl             2.2.0
        tifffile                  2023.4.12
        tiktoken                  0.6.0
        timple                    0.1.8
        tinycss2                  1.2.1
        tornado                   6.4.1
        tqdm                      4.66.1
        traitlets                 5.14.3
        typing_extensions         4.11.0
        typing-inspect            0.9.0
        tzdata                    2023.4
        unicodedata2              15.1.0
        uri-template              1.3.0
        url-normalize             1.4.3
        urllib3                   2.2.1
        virtualenv                20.25.0
        wcwidth                   0.2.5
        webcolors                 1.13
        webencodings              0.5.1
        websocket-client          1.8.0
        websockets                12.0
        wheel                     0.43.0
        win-inet-pton             1.1.0
        xlrd                      2.0.1
        yarl                      1.9.4
[Scheduled Jobs Documentation.docx](https://github.com/user-attachments/files/15917490/Scheduled.Jobs.Documentation.docx)

</pre>
</details>

<details><summary>Command Line Output</summary>
<pre>
Paste the output from your command line running `jupyter lab` here, use `--debug` if possible.
</pre>
</details>

<details><summary>Browser Output</summary>

<pre>
Paste the output from your browser Javascript console here, if applicable.

</pre>
</details>

**Comment by TallGibbs** [comment]
Here is the Anaconda Prompt terminal output when the error occurs trying to schedule a job:
[E 2024-06-20 13:44:47.467 SchedulerApp] module 'fsspec' has no attribute 'open'
    Traceback (most recent call last):
      File "C:\Users\XXXXX\AppData\Local\anaconda3\envs\Projects\Lib\site-packages\jupyter_scheduler\handlers.py", line 127, in post
        self.scheduler.create_job_definition(CreateJobDefinition(**payload))
      File "C:\Users\XXXXX\AppData\Local\anaconda3\envs\Projects\Lib\site-packages\jupyter_scheduler\scheduler.py", line 544, in create_job_definition
        self.copy_input_file(model.input_uri, staging_paths["input"])
      File "C:\Users\XXXXX\AppData\Local\anaconda3\envs\Projects\Lib\site-packages\jupyter_scheduler\scheduler.py", line 370, in copy_input_file
        with fsspec.open(input_filepath) as input_file:
             ^^^^^^^^^^^
    AttributeError: module 'fsspec' has no attribute 'open'
[W 2024-06-20 13:44:47.469 ServerApp] 500 POST /scheduler/job_definitions?1718905487425 (::1): Unexpected error occurred during creation of job definition.
[W 2024-06-20 13:44:47.475 SchedulerApp] wrote error: 'Unexpected error occurred during creation of job definition.'
    Traceback (most recent call last):
      File "C:\Users\XXXXX\AppData\Local\anaconda3\envs\Projects\Lib\site-packages\jupyter_scheduler\handlers.py", line 127, in post
        self.scheduler.create_job_definition(CreateJobDefinition(**payload))
      File "C:\Users\XXXXX\AppData\Local\anaconda3\envs\Projects\Lib\site-packages\jupyter_scheduler\scheduler.py", line 544, in create_job_definition
        self.copy_input_file(model.input_uri, staging_paths["input"])
      File "C:\Users\XXXXX\AppData\Local\anaconda3\envs\Projects\Lib\site-packages\jupyter_scheduler\scheduler.py", line 370, in copy_input_file
        with fsspec.open(input_filepath) as input_file:
             ^^^^^^^^^^^
    AttributeError: module 'fsspec' has no attribute 'open'

    The above exception was the direct cause of the following exception:

    Traceback (most recent call last):
      File "C:\Users\XXXXX\AppData\Local\anaconda3\envs\Projects\Lib\site-packages\tornado\web.py", line 1790, in _execute
        result = await result
                 ^^^^^^^^^^^^
      File "C:\Users\XXXXX\AppData\Local\anaconda3\envs\Projects\Lib\site-packages\jupyter_scheduler\handlers.py", line 140, in post
        raise HTTPError(
    tornado.web.HTTPError: HTTP 500: Internal Server Error (Unexpected error occurred during creation of job definition.)
[E 2024-06-20 13:44:47.478 SchedulerApp] {
      "Host": "localhost:8888",
      "Accept": "*/*",
      "Referer": "http://localhost:8888/lab/tree/SCM%20DB/Policy%20%26%20User%20Guides%20%26%20Updated%20Doc/.DWH/Production%20Ready%20Jupyter%20Notebooks",
      "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/126.0.0.0 Safari/537.36"
    }
[E 2024-06-20 13:44:47.478 SchedulerApp] 500 POST /scheduler/job_definitions?1718905487425 (c3adefb8021548849d573467c4ef4bb5@::1) 51.01ms referer=http://localhost:8888/lab/tree/SCM%20DB/Policy%20%26%20User%20Guides%20%26%20Updated%20Doc/.DWH/Production%20Ready%20Jupyter%20Notebooks

**Comment by andrii-i** [comment]
Hi @TallGibbs. Thank you for creating this issue. Most of our users are not running Windows so feedback from Windows users is especially valuable. Looking into this.

**Comment by andrii-i** [comment]
@TallGibbs thank you for the update. The error message reads `AttributeError: module 'fsspec' has no attribute 'open'` so it seems problem has to do with 'fsspec' dependency in your Windows 10 installation.

Could you please try creating a fresh conda environment, installing JupyterLab and Jupyter Scheduler, and seeing if the problem persists?

**Comment by TallGibbs** [comment]
@andrii-i  

Just to confirm: create the fresh conda environment that is identical to "root", and only add required packages from there.  Correct? Do this on my Windows 10 machine?

Do NOT make a new environment with my .yaml file because that would/should just be identical to what I am using now on my Windows 10 machine.

**Comment by andrii-i** [comment]
@TallGibbs I wanted you to try reproducing the error in a new conda environment on your Windows 10 machine. Please try only installing JupyterLab and Jupyter Scheduler at first and seeing if you can repro the problem. This way it would be clear if problem has to do with .yaml file and combination of packages inside it.

**Comment by TallGibbs** [comment]
Ok I understand and will try that next (probably tomorrow) and will let you know.

**Comment by TallGibbs** [comment]
@andrii-i  
Doing the fresh environment with only pandas (2.2.0), Jupyter Lab and Jupyter Scheduler resolved the issue.  Interestinly, between my original environment and this new environment it still uses the same version of fsspec (2023.6.0), so some other package is modifying that package somehow?

See attached for documentation on this new enviornments "jupyter troubleshoot" output; this is still on Windows 10 but the clean environment which now allows jobs to be scheduled.
[Scheduled Jobs Documentation_v2_schedule_test.docx](https://github.com/user-attachments/files/15975248/Scheduled.Jobs.Documentation_v2_schedule_test.docx)

**Comment by andrii-i** [comment]
@TallGibbs Thank you for the update. I'm happy that Jupyter Scheduler works correctly. It's hard to say what exactly went wrong with fsspec. It also might have to do with fsspec installation.

One thing that I noticed is that in that new environment (no error) uses latest version of jupyter_scheduler 2.7.1 while old one (with error) uses jupyter_scheduler 2.4.0. Maybe updating jupyter_scheduler in your .yaml file to latest version or even removing version pins and letting conda re-resolve dependencies would help?

**Comment by TallGibbs** [comment]
@andrii-i  I did "pip uninstall jupyter_scheduler", then opened a new session of jupyter lab and installed jupyter_scheduler from the "extensions" within jupyter lab.  This installed the newest 2.7.1 version and when I tried to "run now" on a notebook, I got the same error.  So, simply updating this 1 package is not resolving the conflict; there must be conflicts in other packages which got triggered when I did my multiple updates at once back on June 12th.

For now, I'm going to see if I can fully build out my new environment to include ALL packages needed to run my script and see if I get the scheduler to work.
